beta/nncf/tensorflow/batchnorm_adaptation.py

```diff
@@ -11,21 +11,81 @@
 """
 
+from itertools import islice
+
+import tensorflow as tf
+
 from nncf.common.batchnorm_adaptation import BatchnormAdaptationAlgorithmImpl
+from beta.nncf.tensorflow.graph.metatypes.keras_layers import TFBatchNormalizationLayerMetatype
+from beta.nncf.tensorflow.graph.metatypes.matcher import get_keras_layer_metatype
+from nncf.common.utils.progress_bar import ProgressBar
+
+
+class BNTrainingStateSwitcher:
+    """
+    Context manager for switching BatchNormalization layer train mode.
+    At the enter, set BatchNormalization layers to the train mode.
+    At the exit, restore original BatchNormalization layer mode.
+    """
+    def __init__(self, model: tf.keras.Model):
+        self._model = model
+        self._original_training_state = {}
+
+    def __enter__(self):
+        for layer in self._model.layers:
+            if get_keras_layer_metatype(layer) == TFBatchNormalizationLayerMetatype:
+                self._original_training_state[layer] = layer.trainable
+                layer.trainable = True
+
+    def __exit__(self, *exc):
+        for layer in self._model.layers:
+            if get_keras_layer_metatype(layer) == TFBatchNormalizationLayerMetatype:
+                layer.trainable = self._original_training_state[layer]
+
+
+class BNMomentumSwitcher:
+    """
+    Context manager for setting BatchNormalization layer momentum value.
+    At the enter, set BatchNormalization layer momentum to 0.1 to erase the history of statistics.
+    At the exit, restore original BatchNormalization layer momentum (used for further adaptation).
+    """
+    def __init__(self, model: tf.keras.Model):
+        self._model = model
+        self._original_momentum_values = {}
+
+    def __enter__(self):
+        for layer in self._model.layers:
+            if get_keras_layer_metatype(layer) == TFBatchNormalizationLayerMetatype:
+                self._original_momentum_values[layer] = layer.momentum
+                layer.momentum = 0.1
+
+    def __exit__(self, *exc):
+        for layer in self._model.layers:
+            if get_keras_layer_metatype(layer) == TFBatchNormalizationLayerMetatype:
+                layer.momentum = self._original_momentum_values[layer]
 
 
 class TFBatchnormAdaptationAlgorithmImpl(BatchnormAdaptationAlgorithmImpl):
     """
     Implementation of the batch-norm statistics adaptation algorithm for the TensorFlow backend.
     """
-
-    def run(self, model) -> None:
+    def run(self, model: tf.keras.Model) -> None:
         """
         Runs the batch-norm statistics adaptation algorithm.
 
         :param model: A model for which the algorithm will be applied.
         """
-        # TODO(andrey-churkin): Should be implemented.
-
-        raise NotImplementedError('There is no possibility to start the batch-norm statistics adaptation algorithm '
-                                  'for the TensorFlow backend because it is not implemented.')
+        with BNTrainingStateSwitcher(model):
```

Comment:
I think we should throw an exception when `device` is not equal to `None` because we don't support other devices.

Reply:
Agree